### PR TITLE
Track WAL in MANIFEST: update WalMetadata for WAL syncing

### PR DESCRIPTION
**Pull request by cheng-chang**
There are some tricky behaviors related to WAL sync:

- When creating a WAL, the WAL might not be synced, if the WAL directory is not synced, the WAL file's metadata may not even be synced to disk, so during recovery, when listing the WAL directory, the WAL may not even show up.
- During each DB::Write, the WriteOption can control whether the WAL should be synced, so a WAL previously not synced on creation can be synced during Write.

For each `SyncWAL`, we'll track the synced status and the current WAL size. Previously, we only track the WAL size on closing.
During recovery, we check that the on-disk WAL size is >= the last synced size.

So this PR introduces `synced_size` and `closed` to `WalMetadata` for the above design update.

Test Plan:
- updated wal_edit_test
- updated version_edit_test

### CODE REVIEW

**Comment by riversand963** [comment]
> There are some tricky behaviors related to WAL sync:
> 
> * When creating a WAL, the WAL might not be synced, if the WAL directory is not synced, the WAL file's metadata may not even be synced to disk, so during recovery, when listing the WAL directory, the WAL may not even show up.
> * During each DB::Write, the WriteOption can control whether the WAL should be synced, so a WAL previously not synced on creation can be synced during Write.
> * On closing a WAL, if the WAL is not synced, then the size of the WAL may grow afterwards because the some data hasn't been synced to disk.

After closing WAL, how can the size of the WAL keep growing?

> 
> So this PR introduces `synced` and `closed` to `WalMetadata`, then during recovery, we can differentiate among different combinations of `synced` and `closed`.
> 
> Test Plan:
> 
> * updated wal_edit_test
> * updated version_edit_test

**Comment by cheng-chang** [comment]
@riversand963 I mean the WAL is closed, but not synced. So if OS syncs data asynchronously, then the WAL size will grow.

File: db/wal_edit.h [hunk 1]
```diff
@@ -30,6 +32,12 @@ class WalMetadata {
 
   explicit WalMetadata(uint64_t size_bytes) : size_bytes_(size_bytes) {}
 
+  bool IsClosed() const { return HasSize(); }
```
Comment (riversand963) [inline db/wal_edit.h:35]: Can you remind me why `IsClosed()` is equivalent to `HasSize()`?

Comment (cheng-chang) [reply db/wal_edit.h:38]: left inline comments

Comment (riversand963) [reply db/wal_edit.h:38]: If application calls `SyncWAL()` or writes with `WriteOptions.sync=true`, then the WAL will have part of it synced. Recovery should check the actual WAL file size against this.
Things can be more complex if we allow log file number recycling.

Comment (cheng-chang) [reply db/wal_edit.h:38]: We will only write size of the WAL after closing it.

**Review by riversand963 — APPROVED** [review]
LGTM. Thanks @cheng-chang for the PR

**Pull request closed without merging**